lib/vsc/mympirun/pmi/sched.py

```diff
@@ -0,0 +1,180 @@
+            self.log.debug("Keeping existing environment variable %s with value %s (ignore value %s)",
+                           key, current, value)
+        else:
+            os.environ[key] = str(value)
+            self.envs.append(key)
+            self.log.debug("Set environment variable %s: %s", key, value)
+
+    def pmicmd(self):
+        """
+        Return generated pmi command (as list) and the run function
+        envs is list of variable names that is modified
+        """
+        pmicmd = [self.LAUNCHER]
+
+        # the mpi calls should only set environment variables
+        for name in ['tune', 'pmi', 'debug']:
+            method_name = 'mpi_' + name
+            getattr(self, method_name)()
+            self.log.debug("Calling %s", method_name)
+
+        for name in ['sched', 'sizing', 'environment', 'mpi', 'debug']:
+            args = getattr(self, 'pmicmd_' + name)()
+            self.log.debug("Generated pmicmd %s arguments %s", name, args)
+            pmicmd.extend(args)
+
+        run_function, run_function_args = self.run_function()
+        pmicmd.extend(run_function_args)
+
+        pmicmd.extend(['--' + x for x in getattr(self.options, 'pass', [])])  # .pass gives syntax error?
+
+        self.log.debug("Generated pmicmd %s", pmicmd)
+        return pmicmd, run_function
+
+    def run_function(self):
+        """
+        Return required run function and run function related args as list
+        """
+        if self.options.output:
+            def run_output(cmd, **kwargs):
+                kwargs['filename'] = self.options.output
+                return run_file(cmd, **kwargs)
+            return run_output, []
+        else:
+            return async_to_stdout, []
+
+    def pmicmd_sched(self):
+        """Generate the sched related arguments to the launcher as a list"""
+        return []
+
+    def job_info(self, job_info):
+        """
+        Fill in/complete/edit job_info dict and return it
+        """
+        self.log.warn("Nothing done with job_info %s", job_info)
+        return job_info
+
+    def pmicmd_size_args(self, mpi_info):
+        """
+        Convert mpi_info into launcher list of args
+        """
+        self.log.warn("Nothing done with mpi_info %s, no args generated", mpi_info)
+        return []
+
+    def sane_job_info(self, info, hdr='info'):
+        """
+        info dict (e.g. job_info and mpi_info dicts) must match the JOB_INFO template
+        to make sure that after passing around and manipulation no unexpected data appears
+        or expected data disappear.
+        """
+        expected_keys = set(JOB_INFO.keys())
+        info_keys = set(info.keys())
+        if expected_keys == info_keys:
+            self.log.debug("%s is sane", hdr)
+        else:
+            unexpected_keys = info_keys - expected_keys
+            if unexpected_keys:
+                self.log.raiseException("Keys %s only in %s, not in JOB_INFO" % (hdr, sorted(unexpected_keys)))
```

Comment:
order of arguments to format log statement are wrong?

Reply:
i'm going to get rid of this.